### DISCUSSION TITLE
NeoLemmix / SuperLemmix pull request

### DISCUSSION THREAD

**Discussion by Willicious** [graphics32/graphics32#306] [General]
Hi Anders,

Apologies for messaging you via the "Issues", but I don't know any other way to get in touch. I've been trying to contact you regarding the recent NeoLemmix pull request you submitted to Namida on Bitbucket. I develop a fork of NeoLemmix called SuperLemmix, which uses the same version of the GR32 library and is based on the same code for which you submitted the request.

I'm wondering if the updates might also work in my program. I'm happy to share the source with you if you want to take a look. Is it possible for me to merge or cherry pick the updates you submitted? Also, since I've already integrated GR32 into RAD, will I need to do this again to get the latest version (I'm pretty sure I already have the latest version, but I'd have to check)?

If you know of a way to PM me on here, please feel free to do so.

Kind regards,

Will

**Comment by andersmelander** [comment]
Hi Will,

As far as I can tell, SuperLemmix was forked from NeoLemmix end of January last year, so they are not that far apart.

I just tried cherry-picking the commits from [the NeoLemmix PR](https://bitbucket.org/namida42/neolemmixplayer/pull-requests/1) into SuperLemmix and while there were some conflicts on 2-3 of the commits it was nothing that can't easily be resolved. The changes are fairly minor so you can either apply them manually or cherry-pick and resolve the conflicts.

I haven't compiled and run SuperLemmix but I believe all the changes also applies to your fork.

You will need to update your Graphics32. [Some of the changes are for compatibility with Graphics32 v3](https://bitbucket.org/anders_melander/neolemmixplayer/commits/64fbdf4473fb03b71ee7d3df917aa8acd50f8ff7) (that is, the current master branch) and as far as I can see SuperLemmix, like NeoLemmix, is only compatible with v1.9 and v2.0.

**Reply by Willicious** [reply]
Hi Anders,

Thanks for your swift reply and for checking the SuperLemmix code! SLX is by now very different from NL on the surface, but the basic codebase remains pretty much the same.

How does upgrading GR32 work? I remember it was a bit of a kerfuffle to get the NL source to compile, and it involved manually adding the GR32 libraries/dependencies.

Would it be best to first update GR32 and then cherry pick the NL PR commits, or vice versa? Or does it not matter? If SLX no longer compiles once GR32 has been updated (very possible), is it easy enough to roll back to the previous version?

Apologies for the noobish questions, this is really my first major project and I'm still getting used to how everything works. Many thanks for your help with this 😊

**Reply by andersmelander** [reply]
> Would it be best to first update GR32 and then cherry pick the NL PR commits, or vice versa?

Start by updating Graphics32. Create a branch in SuperLemmix and cherry pick the commits into that branch. You can recompile after each cherry pick or you can wait to you have them all.
Actually, if you add me as a contributor to your repository then I can push the branch to you as I already have all the changes applied to a branch in my local copy.

> If SLX no longer compiles once GR32 has been updated (very possible), is it easy enough to roll back to the previous version?

Yes, it's easy enough. Just restore the old Graphics32 (you did remember to save a copy, right? ;-) ), and switch back to the master branch.

By accident I just stumbled across this old thread: https://en.delphipraxis.net/topic/8878-change-fadeout-code-to-fadein-code/
I now realize that all your questions at DelphiPraxis had to do with SuperLemmix. The fading you mention in the above thread is actually handled in the pull request by this commit: https://bitbucket.org/anders_melander/neolemmixplayer/commits/ac66fc6a6f3f8a0cdb628a323ef31ee004c7d1b8

**Reply by Willicious** [reply]
Brilliant, thanks for this! And yes, I was sure I recognised your name from somewhere! 😊

I would like to add you as a contributor, it's asking for user's email address which doesn't seem right:

![image](https://github.com/graphics32/graphics32/assets/76848857/ea831905-b3d7-4220-9a08-9aa72e080dc6)

**Comment by andersmelander** [comment]
Try **andersmelander**

**Comment by andersmelander** [comment]
The other is my bitbucket username :-)

**Comment by Willicious** [comment]
Tried that, still getting no results. Then again, I don't seem to be getting _any_ results at the moment, not sure why. Is there any other way of inviting you?

**Comment by Willicious** [comment]
I've created a topic on the Atlassian community, hopefully someone there might know what I'm doing wrong:
https://community.atlassian.com/t5/Bitbucket-questions/How-do-I-Invite-a-user-to-my-repository/qaq-p/2676126#M103505

Bear with me, we'll get there eventually! :-)

**Reply by andersmelander** [reply]
Huh? Your SuperLemmix repo is on Github so you need to add me as a collaborator on Github using my Github username.
Why are you doing it on bitbucket?

**Reply by Willicious** [reply]
Because I use GitHub as a backup. The actual working copy is on Bitbucket, here:
https://bitbucket.org/Will_James/superlemmixplayer/src/master/

Nobody has replied on Atlassian yet so I'm still not sure how to add you, but maybe you can send the same pull request to the SLX Bitbucket repo and I can add you from there? Not sure how it works, thanks for your patience with this 👍

**Reply by Willicious** [reply]
Meanwhile, I have added you to the GitHub repo so perhaps we can work from there. I've just pushed the latest changes so both repos are identical - I won't commit anything else until we're done with updating GR32 - _really_ appreciate your help with this, the codebase definitely needs a graphics update ASAP

**Reply by andersmelander** [reply]
Don't expect _any_ response from Atlassian. I would advise you to migrate to Github or Gitlab as Bitbucket is becoming more and more broken.

Anyway, I've now pushed my branch to the repo on Github. If you have your local repo configured with both Bitbucket and Github as remotes then you can just pull the branch to your local repo from Github.

**Reply by Willicious** [reply]
Yes, tbh I would much prefer to do everything via GitHub. At present, my local repo is managed through SourceTree and pushes to the Bitbucket repo.

I do like the SourceTree interface so I'd be happy to continue working with that, but hopefully there's a way to sync the local repo with GitHub instead (i.e. from within SourceTree). Now that both repos are identical, I'll look into doing that and then I can go ahead and accept your branch.

First, though, I'll need to update GR32 and make sure SLX still compiles; any info you can give to help with this is much appreciated. Currently, I have GR32 installed in RAD Studio; I followed these instructions to get it installed (note that this is a post I made on the Forums to help others compile the source, but also so I could remember how to do it!):

https://www.lemmingsforums.net/index.php?topic=6231.msg98527#msg98527

Do I just need to swap the existing GR32 files for the new ones, or will I need to re-install it into RAD?

**Reply by andersmelander** [reply]
> hopefully there's a way to sync the local repo with GitHub instead (i.e. from within SourceTree).

Just add all the different remote repos as "remotes" in SourceTree:
![billede](https://github.com/graphics32/graphics32/assets/18141921/8e589a16-f526-4b6f-b82f-d4f5a36e5291)
![billede](https://github.com/graphics32/graphics32/assets/18141921/afb0b7d4-141d-41e4-a9b1-16f1bed0440b)
![billede](https://github.com/graphics32/graphics32/assets/18141921/a517f9de-5851-4f01-9b5c-0345484a051a)

and now you will be able to see their branches and commits in the tree:
![billede](https://github.com/graphics32/graphics32/assets/18141921/fb49fa33-b447-4f7c-b645-4a631934ddc7)

**Reply by andersmelander** [reply]
> Do I just need to swap the existing GR32 files for the new ones, or will I need to re-install it into RAD?

1. Make a clone onto your local system of Graphics32 from https://github.com/graphics32/graphics32.git
  It doesn't matter where you put it as you will reference it via the Lemmix project options in Delphi.
2. Open ...\Graphics32\Source\Packages\D104\GR32.groupproj in Delphi
  Replace D104 with whatever version of Delphi you are using.
  For Delphi 11 and later it's called Graphics32_ packages.groupproj; I should get that fixed.
3. Compile **GR32_R**.
4. Compile **GR32_D** and install.
5. Open your Lemmix project and add the Graphics32\Source to the library search path.
  You shouldn't commit this change to the repo as it will be different for everyone.
  You can also add it to Delphi's global library search path instead. I just prefer not to mess with that.
7. Recompile Lemmix.

**Reply by Willicious** [reply]
Great, thanks for these instructions! I have successfully added the GitHub repo as a remote in SourceTree so I can see all of the commits including yours! This is good, I should be able to de-conflict and merge them all in no problem.

To get GR32v3 installed, I first had to manually uninstall the XE8 version of GR32_D to be able to install the new one, but it all seems to have worked. I haven't merged the changes from your commits yet but I will do that later this evening and see how it goes, this will ultimately show whether or not I've done everything correctly.

For now, do you know how to add the GR32 source to the project's library path (i.e. instead of Delphi's global path)? I can only find the global libraries option, there doesn't seem to be one in project options.

**Reply by andersmelander** [reply]
![billede](https://github.com/graphics32/graphics32/assets/18141921/ce7105a5-05c9-45f7-9983-d8866ef1de47)

**Reply by Willicious** [reply]
OK, I kind of knew this would happen but everything is a huge mess and I now don't know how to undo it. When I first added GitHub as a remote, nothing was showing up in the History so I clicked "Pull" and then manually pulled everything from the GitHub branch. This placed all of your commits onto the master branch ready to be de-conflicted/merged, but I wasn't ready for that yet because I hadn't updated GR32, so I unstaged them all.

Then, having updated GR32, I tried to once again stage your commits, but it wouldn't happen. I tried merging them and nothing happened. So, I tried deleting the GitHub remote from the project configuration and now I'm looking at this:

![image](https://github.com/graphics32/graphics32/assets/76848857/f7095976-b3b8-41e6-b547-cb50c69a7a84)

As you can see, all of your commits are there but I don't know why. None of the changes you've made are showing up in my local project. Merging doesn't work. I now have a long red line which appears to be all of the GitHub commits, but I don't know why they're there or how to get rid of them. SuperLemmix will no longer compile. Please help!

**Reply by Willicious** [reply]
Git branches are far too complex for me at the moment, I'm just going to manually add all of the changed files and see if that resolves the errors

**Reply by andersmelander** [reply]
First, let's clean up. Don't panic. Instructions coming....

**Reply by andersmelander** [reply]
No matter what you do: **Don't push anything**. Even if you mess up your local repo the remote will still be fine.

1. Select you **master** branch.
2. Right click the last good commit in the master branch.
  That'll be commit ca3bf19d94f7e2418423080957eda51644768fec with the message "==== Version 2.7.3 ===="
3. From the popup menu select **reset master to here**.
4. Now you local copy of master should like before you started.
5. If you are satisfied that master is now OK, select **Push** and check the **Force push** option.
6. If Force push isn't enabled then you need to go into the SourceTree options, under the Git tab, and enable it.
7. Once you've force pushed, the remote repo will now match your local.

**Reply by andersmelander** [reply]
As far as I can see the problem is that your Github repo isn't a clone of the Bitbucket repo; They are completely independant repos with completely independent commits.  This is why merging a branch from the one repo to the other will make a mess.

My guess is that your have kept them in sync by simply copying files manually. This is definitely not the way to do it (as you've just experienced) but that's another story.

**Reply by andersmelander** [reply]
I can see that you've just gone ahead and added the changes. I would still recommend that you do a reset and force push to go back to before the merge or this will follow you (and everybody why clones the repo) going forward.

**Reply by andersmelander** [reply]
In the future:

- If you f*ck up, don't push. Fix the problems in the local repo before you push anything,
- Don't merge until things has been tested.
- If you don't understand merge, use cherry-picking instead. Cherry-picking copies commits and can be used to integrate changes from other branches, or even other repos (as is your case).
- Don't commit into master and then push until you have actually tested the changes.

And this has nothing to do with Git, but you've just added the path of the compiled units from the run- and design-time packages to the project search path. It should have been the path to the source files. On top of that the path you added is your personal local path which will not be valid for anyone else.

Finally, regarding this _"note that line 183 might need to be what's currently commented out on line 184"_, neither is correct.
The following is most likely the correct solution:

```delphi
TempBmp := TBitmap32.Create;
try
  TempBmp.Assign(Bmp);
  RemoveAlpha(TempBmp);
  PNGWriter.SaveToStream(TempBmp, aStream);
finally
  TempBmp.Free;
end;
```

**Reply by Willicious** [reply]
The good news is that SuperLemmix _does_ compile after the changes have been made! And the screen Fadeout does look much smoother - this is great 👍

In any case, I'll follow your advice and go ahead and reset to before all of this happened. Please bear with me because I'll then need your help with adding your commits again; I don't understand how to merge or cherry pick. If you don't mind helping me with it, it will be invaluable stuff to know going forward...

**Reply by Willicious** [reply]
Which option should I choose? I'm guessing that soft is best here, since we ultimately want to keep the changes... I can make the necessary corrections and then push the changes again, perhaps...?

![image](https://github.com/graphics32/graphics32/assets/76848857/263c1f37-5774-4184-a4fc-cffefce22360)

Also, not sure what you mean about the project search path. Exactly how should that look? I realise it has to have "kernel;" but then I'm not sure what should be after that. Previously, this was set automatically (presumably because I'd set the library path globally...?)

**Reply by andersmelander** [reply]
Hard

**Reply by andersmelander** [reply]
> Also, not sure what you mean about the project search path. Exactly how should that look? I realise it has to have kernel; but then I'm not sure what should be after that. Previously, this was set automatically (presumably because I'd set the library path globally...?)

Let's deal with that later

**Reply by Willicious** [reply]
OK, gone ahead and done a hard reset. This is what we're looking at - the branch with your commits on it is still there (not sure how!):

![image](https://github.com/graphics32/graphics32/assets/76848857/05c040a3-107d-4c28-916e-e0f31bb2c61a)

**Reply by andersmelander** [reply]
> the branch with your commits on it is still there (not sure how!):

Because you pushed it. It will disappear in a moment.

Now you need to do a **force push**.

If you press the **Push** button, is the **Force Push** option enabled?
![billede](https://github.com/graphics32/graphics32/assets/18141921/eb29137a-9048-40da-845c-f63e7124d8e0)

**Reply by andersmelander** [reply]
If not, go into Tools, Options and enable it:
![billede](https://github.com/graphics32/graphics32/assets/18141921/c08a9e8e-90d7-4a6f-895a-7db1da0fce67)

**Reply by Willicious** [reply]
Yes, it is. Before I do that though, SLX is back to not compiling due to GR32 being updated but the necessary changes not yet being integrated into SLX... I'm guessing, though, that this isn't a problem repo-side. Shall I go ahead and force push?

**Reply by andersmelander** [reply]
Yes

**Reply by andersmelander** [reply]
Okay, now add Github as a remote again

**Reply by Willicious** [reply]
Success! The project is restored to how it was before I f**ked it up!

**Reply by Willicious** [reply]
Done, but nothing is showing from it yet:

![image](https://github.com/graphics32/graphics32/assets/76848857/a0bc7f2d-3fa0-42ce-b277-08091be1f7eb)

**Reply by Willicious** [reply]
> As far as I can see the problem is that your Github repo isn't a clone of the Bitbucket repo; They are completely independant repos with completely independent commits. This is why merging a branch from the one repo to the other will make a mess.
> 
> My guess is that your have kept them in sync by simply copying files manually. This is definitely not the way to do it (as you've just experienced) but that's another story.



You're right on with this btw, that's exactly how I've been managing it. I need to step my game up

**Reply by andersmelander** [reply]
Do a Fetch.

With master (Bitbucket) active, right click each of the commits from the GR32v3 branch on Github and select cherry-pick.
![billede](https://github.com/graphics32/graphics32/assets/18141921/f7932d3c-00b4-43d9-9012-659aeaf44694)
 (I'm using another Git client)

**Reply by andersmelander** [reply]
I have the wrong master selected in the above screenshow

**Reply by andersmelander** [reply]
![billede](https://github.com/graphics32/graphics32/assets/18141921/195a5831-b7d1-4421-9fc5-feee860b853b)

**Comment by andersmelander** [comment]
Once you have cherry-picked all the commits, you can add the Graphics32 source folder to your global library search path, compile and run.
Only once you are satisfied that everything is fine do you push the changes.

**Reply by Willicious** [reply]
Should I leave these checked? I have Sourcetree set to auto-push on commit, so I'm guessing I should probably uncheck these?

![image](https://github.com/graphics32/graphics32/assets/76848857/8c83992f-9f0d-4a46-900c-36688f2ba8f4)

**Reply by andersmelander** [reply]
1. ~Don't~ Never auto-push on commit
2. Both options are fine.

**Reply by Willicious** [reply]
Just wanting to double-check that this is OK:

```
class procedure TPngInterface.SavePngStream(aStream: TStream; Bmp: TBitmap32; NoAlpha: Boolean = false);
var
  TempBmp: TBitmap32;
  PNGWriter: IImageFormatWriter;
begin
  PNGWriter := ImageFormatManager.Writers.FindWriter('.png');
  Assert(PNGWriter <> nil);

  if not NoAlpha then
    PNGWriter.SaveToStream(Bmp, aStream)
  else begin
    TempBmp := TBitmap32.Create;
    try
      TempBmp.Assign(Bmp);
      RemoveAlpha(Bmp);
      PNGWriter.SaveToStream(Bmp, aStream)
    finally
      TempBmp.Free;
    end;
  end;
end;
```

Also, I had to make a correction here, just double checking that these are OK (they now match the declarations)):

![image](https://github.com/graphics32/graphics32/assets/76848857/c1f3204d-4fca-487d-8776-01f8409bba14)

**Reply by andersmelander** [reply]
```delphi
TempBmp := TBitmap32.Create;
try
  TempBmp.Assign(Bmp);
  RemoveAlpha(TempBmp);
  PNGWriter.SaveToStream(TempBmp, aStream);
finally
  TempBmp.Free;
end;
```

The rest looks okay.

**Reply by Willicious** [reply]
The project now compiles just fine, and the Fadeout looks great. Thanks so much for you help 👍👍

**Reply by Willicious** [reply]
This has reminded me - you mentioned that it might now be possible to write a FadeIn procedure now that GR32 has been updated...?

**Reply by andersmelander** [reply]
>  you mentioned that it might now be possible to write a FadeIn procedure now that GR32 has been updated...?

Did I? Well, you can just use the exact same method I have used to fade-out.

**Reply by Willicious** [reply]
I see. It looks like you're using the alpha channel, which is good. Is it this line that needs to be reversed?

`ScreenImg.Bitmap.MasterAlpha := MulDiv(255, RemainingTime, MAX_TIME);`

EDIT: I'll reply in [the DelphiPraxis topic](https://en.delphipraxis.net/topic/8878-change-fadeout-code-to-fadein-code/). I'm getting the same issues I have every time I attempt to get anywhere with it, but it seems appropriate to continue this particular discussion there.

**Reply by andersmelander** [reply]
yes

**Reply by Willicious** [reply]
Thanks, I've replied in-topic.

Also, thanks again for walking me though the above steps - this will come in very handy. I've only been coding for about a year and a half and have never had any formal training; I've relied mainly on online help and the occasional tidbits from namida. I've achieved a lot with SuperLemmix code-side but I'm sure I haven't really been doing things properly project-management-side. I'll know what to do now going forward 👍

Next step is probably to try and base the project in GitHub rather than Bitbucket, but... that's for another time!

**Comment by Willicious** [comment]
Hi Anders, just noticed the following hints on compile. It's wherever "ScreenImg.BeginUpdate" is called in the following units:

![image](https://github.com/graphics32/graphics32/assets/76848857/47d1ac40-cbf0-43b8-aeab-3f8c355b907c)

**Reply by andersmelander** [reply]
Those are standard inlining hints. You can add GR32_Image to the uses list in the implementation section if they bother you but there will probably be others.

**Comment by Willicious** [comment]
By the way, on [this page](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) it says that in order to create a PR, it must be on an existing branch. But, you created a new branch for your PR. Just wondering how you did it.

**Reply by andersmelander** [reply]
I can't see where it says that.

Here's what I did (which is how it's normally done):

1. Created a fork of the repo
2. Cloned that fork to my local system.
3. Created a branch on my local clone.
4. Committed the changes into the branch.
5. Pushed the branch to my fork.
6. Created a PR at the original repo to have my branch at my fork merged into the master branch at the original repo.

If I had write access to the original repo I could instead have:
1. Cloned the repo to my local system.
2. Created a branch on my local clone.
3. Committed the changes into the branch.
4. Pushed the branch to the repo.
5. Created a PR at the repo to have my branch merged into the master branch.

You can try the above on your own repo.

**Comment by Willicious** [comment]
Apologies for the formatting of this message, the <>/`` for code isn't working. Found a bug with the GR32v3 update today in PngInterface.Pas line 168, the updated SaveToStream procedure:

```
class procedure TPngInterface.SavePngStream(aStream: TStream; Bmp: TBitmap32; NoAlpha: Boolean = false);
var
  TempBmp: TBitmap32;
  PNGWriter: IImageFormatWriter;
begin
  PNGWriter := ImageFormatManager.Writers.FindWriter('.png');
  Assert(PNGWriter <> nil);

  if not NoAlpha then
    PNGWriter.SaveToStream(Bmp, aStream)
  else begin
    TempBmp := TBitmap32.Create;
    try
      TempBmp.Assign(Bmp);
      RemoveAlpha(Bmp);
      PNGWriter.SaveToStream(Bmp, aStream)
    finally
      TempBmp.Free;
    end;
  end;
end;
```

The following procedure in NNeoLemmixLevelSelect calls SavePngFile, which in turn calls SavePngStream:

```
procedure TFLevelSelect.btnSaveImageClick(Sender: TObject);
var
  N: TTreeNode;
  Obj: TObject;

  BasePack: TNeoLevelGroup;
  PathString: String;

  BMP: TBitmap32;
  SaveDlg: TSaveDialog;
begin
  N := tvLevelSelect.Selected;
  if N = nil then Exit;

  Obj := TObject(N.Data);

  if Obj is TNeoLevelGroup then
  begin
    DumpImagesFallbackFlag := false;
    BasePack := TNeoLevelGroup(Obj).ParentBasePack;

    PathString := MakeSafeForFilename(BasePack.Name);

    BasePack.DumpImages(AppPath + 'Dump\' + PathString + '\');
    {$ifdef exp}
    BasePack.DumpNeoLemmixWebsiteMetaInfo(AppPath + 'Dump\' + PathString + '\');
    {$endif}

    if DumpImagesFallbackFlag then
      ShowMessage('Some styles used in this group appear to be missing. Use the Style Manager to download these. Level images with fallbacks saved to "Dump\' + PathString + '"')
    else
      ShowMessage('Level images saved to "Dump\' + PathString + '"');
  end else if Obj is TNeoLevelEntry then
  begin
    BMP := TBitmap32.Create;
    SaveDlg := TSaveDialog.Create(self);
    try
      if GameParams.Level.HasAnyFallbacks then
        ShowMessage('Some styles used by this level appear to be missing. Use the Style Manager to download these.');
      SaveDlg.Options := [ofOverwritePrompt];
      SaveDlg.Filter := 'PNG File|*.png';
      SaveDlg.DefaultExt := '.png';
      if SaveDlg.Execute then
      begin
        GameParams.Renderer.RenderWorld(BMP, true);
        TPngInterface.SavePngFile(SaveDlg.FileName, BMP);
      end;
    finally
      BMP.Free;
    end;
  end else
    Exit;
end;

class procedure TPngInterface.SavePngFile(fn: String; Bmp: TBitmap32; NoAlpha: Boolean = false);
var
  TempStream: TFileStream;
begin
  TempStream := TFileStream.Create(fn, fmCreate);
  try
    SavePngStream(TempStream, Bmp, NoAlpha);
  finally
    TempStream.Free;
  end;
end;
```

We get the PNGWriter assertion (line 174 PngInterface.Pas). Bug is present in both NL and SLX. Not sure what to try, thought you might be able to spot a fix more quickly.

**Comment by andersmelander** [comment]
This is wrong (but not the cause of your problem):

```
    TempBmp := TBitmap32.Create;
    try
      TempBmp.Assign(Bmp);
      RemoveAlpha(Bmp);
      PNGWriter.SaveToStream(Bmp, aStream)
    finally
      TempBmp.Free;
    end;
```

It should be:
```
    TempBmp := TBitmap32.Create;
    try
      TempBmp.Assign(Bmp);
      RemoveAlpha(TempBmp);
      PNGWriter.SaveToStream(TempBmp, aStream)
    finally
      TempBmp.Free;
    end;
```
----

The assertion means that the PNG reader/writer hasn't been registered. To solve that simply add `GR32.ImageFormats.PNG32,` to the uses clause of the unit. I have no idea how that managed to get past me. I must have forgotten to commit something.

Apart from that, all `TPngInterface.SavePngFile(Filename, Bitmap)` could be replaced by `Bitmap.SaveToFile(Filename)` since TBitmap32 automatically saves to PNG if the filetype is '.png'. Same with `TBitmap32.LoadFromFile`.

**Reply by Willicious** [reply]
Replaced the TempBmp stuff, thanks. I forgot to add that particular fix in my working copy of NL.

Meanwhile, tried adding `GR32.ImageFormats.PNG32,` to all units that call TPngInterface.SavePngFile. Still getting the same error unfortunately. Maybe I should try as you've suggested and call Bitmap.SaveToFile instead (only questions there would be... (a) how do I know the filetype will automatically be '.png' - i.e. from code-side? and (b) if going down this route, does NL still need SavePngStream at all?)

**Reply by andersmelander** [reply]
You only need to add the unit once and it doesn't matter where you do it. The file format registration is done in the initialization section so you just need to ensure that it's linked into the application.
Actually, I can see now that you don't need it at all; It's already included by default via the `GR32.ImageFormats.Default` unit which is included in GR32.pas.

There must be something else going on. I think I will have to try and reproduce the problem.

> (a) how do I know the filetype will automatically be '.png'

I don't know if you can know that but it would be strange to use another filetype for PNG.
For SaveToFile the image format is determined by the filetype - with fallback to bmp AFAIR. For LoadFromFile I actually believe that I look at the file signature. Let me check... Yes, LoadFromFile first tries to find a reader based on the filetype and the based on the signature.

> (b) if going down this route, does NL still need SavePngStream at all?)

SavePngStream has a parameter that tells if the PNG should be saved without Alpha. If this options isn't used anywhere then you don't need it.

**Reply by Willicious** [reply]
> You only need to add the unit once and it doesn't matter where you do it. The file format registration is done in the initialization section so you just need to ensure that it's linked into the application. Actually, I can see now that you don't need it at all; It's already included by default via the `GR32.ImageFormats.Default` unit which is included in GR32.pas.

I meant that I added `GR32.ImageFormats.PNG32` to the uses clause of all units that call TPngInterface.SaveToFile. Doing so hasn't solved the issue. I went ahead and tried using Bitmap.SaveToFile instead and this works, but let's try to resolve the GR32 saving anyway, if only because it might highlight something useful elsewhere in each of our projects.

**Reply by andersmelander** [reply]
> I meant that I added GR32.ImageFormats.PNG32 to the uses clause of all units that call TPngInterface.SaveToFile.

Yes, and I meant that you don't need to do that.

**Comment by andersmelander** [comment]
Fixed: https://bitbucket.org/namida42/neolemmixplayer/pull-requests/3
